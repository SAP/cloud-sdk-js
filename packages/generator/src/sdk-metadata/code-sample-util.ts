import { getLevenshteinClosest } from '@sap-cloud-sdk/generator-common/internal';
import { VdmEntity, VdmOperation, VdmParameter } from '../vdm-types';

/**
 * @internal
 */
export function getODataEntity(
  serviceName: string,
  vdmEntities: VdmEntity[]
): VdmEntity {
  return (
    getLevenshteinClosest(serviceName, vdmEntities, x => x.className) ||
    getShortestNameEntity(vdmEntities)
  );
}
/**
 * @internal
 */
export function getShortestNameEntity(vdmEntities: VdmEntity[]): VdmEntity {
  // If no closest entity found, return the entity with shortest name
  return vdmEntities.sort((a, b) =>
    a.className.length < b.className.length ? -1 : 1
  )[0];
}
/**
 * @internal
 */
export function sampleOperationImport(
  serviceName: string,
  operationImports: VdmOperation[]
): VdmOperation {
  if (operationImports.length === 1) {
    return operationImports[0];
  }

  return (
    getLevenshteinClosest(serviceName, operationImports, x => x.name) ||
    getOperationWithoutParameters(operationImports) ||
    getOperationWithMinParameters(operationImports)
  );
}
/**
 * @internal
 */
export function getOperationWithoutParameters(
  operationImports: VdmOperation[]
): VdmOperation | undefined {
  return operationImports.find(func => func.parameters?.length === 0);
}

/**
 * Sorts and gets a function or action import having minimum input parameters.
 * @param operationImports - function or action imports array
 * @returns Function or action containing minimum input parameters
 * @internal
 */
export function getOperationWithMinParameters(
  operationImports: VdmOperation[]
): VdmOperation {
  const getOperations = operationImports.filter(
    func => func.httpMethod?.toLowerCase() === 'get'
  );
  if (getOperations.length) {
    operationImports = getOperations;
  }
  const sortedOperations = operationImports.sort((funcA, funcB) =>
    funcA.parameters?.length < funcB.parameters?.length ? -1 : 1
  );
  return sortedOperations[0];
}
/**
 * @internal
 */
export function getOperationParams(parameters: VdmParameter[]): string {
  const paramString = parameters
    .slice(0, 2)
    .reduce(
      (cumulator, currentParam) =>
        `${cumulator}, ${currentParam.parameterName}: '${currentParam.parameterName}'`,
      ''
    );
  return `{${paramString.substring(1)}${
    parameters.length > 2 ? ', ...' : ''
  } }`;
}
