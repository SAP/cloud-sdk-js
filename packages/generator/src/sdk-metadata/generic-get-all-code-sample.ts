import { TestEntity } from '@sap-cloud-sdk/test-services/v2/test-service';

const result = TestEntity.requestBuilder().getAll().top(5).execute({destinationName:'myDestinationName'})
