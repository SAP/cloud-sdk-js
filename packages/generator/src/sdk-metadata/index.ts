export * from './code-sample-util';
export * from './code-samples';
export * from './generation-and-usage';
export * from './sdk-metadata';
