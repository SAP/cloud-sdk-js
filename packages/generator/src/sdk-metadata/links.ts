import { Links } from './sdk-metadata-types';

export function getLinks(): Links {
  return {
    sdkDocumentation: 'https://sap.github.io/cloud-sdk/',
    featureDocumentation:
      'https://sap.github.io/cloud-sdk/docs/js/features/odata/use-odata-v2-type-safe-client-for-javascript-typescript',
    support: 'https://github.com/SAP/cloud-sdk-js/issues/new/choose',
    apiHubTutorial:
      'https://sap.github.io/cloud-sdk/docs/js/features/odata/execute-odata-request',
    generationManual:
      'https://sap.github.io/cloud-sdk/docs/js/features/odata/generate-odata-client'
  };
}
