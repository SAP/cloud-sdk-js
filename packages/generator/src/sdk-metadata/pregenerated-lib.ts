import { readFileSync } from 'fs';
import { resolve } from 'path';
import { executeHttpRequest } from '@sap-cloud-sdk/core';
import { VdmServiceMetadata } from '../vdm-types';
import { GeneratorOptions } from '../generator-options';
import {
  genericDescription,
  s4hanaCloudDescription
} from '../package-description';
import { PregeneratedLibrary } from './sdk-metadata-types';

export function getInstallationSnippet(service: VdmServiceMetadata): string {
  return `npm i ${service.npmPackageName}:latest`;
}

export function getRepositoryLink(service: VdmServiceMetadata): string {
  return `https://www.npmjs.com/package/${service.npmPackageName}`;
}

export function getGeneratorVersion(): string {
  return JSON.parse(
    readFileSync(resolve(__dirname, '../../package.json'), 'utf8')
  ).version;
}

export function getVersionForClient(options: GeneratorOptions): string {
  return options.versionInPackageJson || getGeneratorVersion();
}

export function getServiceDescription(
  service: VdmServiceMetadata,
  options: GeneratorOptions
): string {
  return options.s4hanaCloud
    ? s4hanaCloudDescription(service.directoryName)
    : genericDescription(service.directoryName);
}

export function getTimeStamp(): string {
  return `/Date(${Date.now()})/`;
}

export async function isPublishedNpmPackage(
  service: VdmServiceMetadata
): Promise<boolean> {
  try {
    await executeHttpRequest(
      { url: 'http://registry.npmjs.org/' },
      { method: 'get', url: service.npmPackageName }
    );
    return true;
  } catch (e) {
    return false;
  }
}

export async function getPregeneratedLibrary(
  service: VdmServiceMetadata,
  options: GeneratorOptions
): Promise<PregeneratedLibrary | undefined> {
  if (await isPublishedNpmPackage(service)) {
    return {
      repository: 'npm',
      dependencyName: service.npmPackageName,
      installLibrarySnippet: getInstallationSnippet(service),
      repositoryLink: getRepositoryLink(service),
      compatibilityNotes: '',
      description: getServiceDescription(service, options),
      generatedAt: getTimeStamp(),
      version: getVersionForClient(options)
    };
  }
}
