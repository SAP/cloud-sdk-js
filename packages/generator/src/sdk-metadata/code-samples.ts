import { codeBlock } from '@sap-cloud-sdk/util';
import { MultiLineText } from '@sap-cloud-sdk/generator-common/internal';
import voca from 'voca';
import { VdmOperation } from '../vdm-types';
import { getApiName } from '../generator-without-ts-morph/service';
import { getOperationParams } from './code-sample-util';

/**
 * @internal
 */
export function entityCodeSample(
  entityName: string,
  serviceName: string,
  directoryName: string
): MultiLineText {
  return codeBlock`
import { ${voca.decapitalize(
    serviceName
  )} } from './generated/${directoryName}';

const { ${getApiName(entityName)} } = ${voca.decapitalize(serviceName)}();
const resultPromise = ${getApiName(
    entityName
  )}.requestBuilder().getAll().top(5).execute({ destinationName: 'myDestinationName' });

`;
}

/**
 * @internal
 */
export function functionImportCodeSample(
  functionImport: VdmOperation,
  packageName: string
): MultiLineText {
  return importsCodeSample(functionImport, packageName);
}
/**
 * @internal
 */
export function actionImportCodeSample(
  actionImport: VdmOperation,
  packageName: string
): MultiLineText {
  return importsCodeSample(actionImport, packageName);
}

function importsCodeSample(
  actionFunctionImport: VdmOperation,
  packageName: string
): MultiLineText {
  return codeBlock`
import { ${actionFunctionImport.name}${
    actionFunctionImport.parametersTypeName
      ? `, ${actionFunctionImport.parametersTypeName}`
      : ''
  } } from './generated/${packageName}';

${getParameterCodeSample(actionFunctionImport)}
const resultPromise = ${actionFunctionImport.name}(${
    actionFunctionImport.parametersTypeName ? 'parameter' : ''
  }).execute({ destinationName: 'myDestinationName' });
`;
}

function getParameterCodeSample(functionImport: VdmOperation): string {
  if (functionImport.parameters) {
    return `const parameter: ${
      functionImport.parametersTypeName
    } = ${getOperationParams(functionImport.parameters)};
    `;
  }
  return '';
}
