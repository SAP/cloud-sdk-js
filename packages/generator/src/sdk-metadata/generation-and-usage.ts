import { MultiLineText } from '@sap-cloud-sdk/generator-common/internal';
import { VdmServiceMetadata } from '../vdm-types';
import {
  actionImportCodeSample,
  entityCodeSample,
  functionImportCodeSample
} from './code-samples';
import { sampleOperationImport, getODataEntity } from './code-sample-util';

/**
 * @internal
 */
export function getApiSpecificUsage(
  service: VdmServiceMetadata
): MultiLineText {
  if (service.entities?.length > 0) {
    const entity = getODataEntity(service.originalFileName, service.entities);
    return entityCodeSample(
      entity.className,
      service.className,
      service.directoryName
    );
  }
  // Return function/action import usage if no entity is found.
  if (service.functionImports?.length > 0) {
    const functionImport = sampleOperationImport(
      service.originalFileName,
      service.functionImports
    );
    return functionImportCodeSample(
      functionImport,
      `${service.directoryName}/function-imports`
    );
  }
  if (service.actionImports) {
    const actionImport = sampleOperationImport(
      service.originalFileName,
      service.actionImports
    );
    return actionImportCodeSample(
      actionImport,
      `${service.directoryName}/action-imports`
    );
  }
  return '';
}
