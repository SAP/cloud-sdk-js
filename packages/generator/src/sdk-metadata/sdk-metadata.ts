import {
  getSdkMetadataClient,
  Client
} from '@sap-cloud-sdk/generator-common/internal';
import { VdmServiceMetadata } from '../vdm-types';
import { getApiSpecificUsage } from './generation-and-usage';

/**
 * @internal
 */
export async function sdkMetadata(
  service: VdmServiceMetadata
): Promise<Client> {
  const generationAndUsage = await getApiSpecificUsage(service);
  return getSdkMetadataClient(generationAndUsage, 'OData');
}
