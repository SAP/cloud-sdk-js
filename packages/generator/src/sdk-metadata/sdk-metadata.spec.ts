import { getTestService } from './pregenerated-lib.spec';
import { sdkMetadata } from './sdk-metadata';

describe('sdk-metadata', () => {
  const service = getTestService();

  it('generates the sdk metadata content for services without pregenerated lib', async () => {
    const metadata = await sdkMetadata(service);
    expect(metadata).toMatchSnapshot({ generatorVersion: expect.any(String) });
  });
});
