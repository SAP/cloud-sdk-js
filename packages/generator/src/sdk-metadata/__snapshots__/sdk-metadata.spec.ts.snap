// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pregenerated-lib returns description of the service 1`] = `"SAP Cloud SDK for JavaScript: Generated client for service Test Service"`;

exports[`sdk-metadata generates the sdk metadata content for services without pregenerated lib 1`] = `
Object {
  "apiSpecificUsage": "import { testService } from './generated/test-service';

const { caseTestApi } = testService();
const resultPromise = caseTestApi.requestBuilder().getAll().top(5).execute({ destinationName: 'myDestinationName' });
",
  "apiType": "OData",
  "generatorVersion": Any<String>,
  "language": "JavaScript",
  "serviceStatus": "verified",
}
`;
